tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+#  (PART) Conclusion {-}
+
+# Tradeoffs
+
+::: {.alert .alert-success} 
+This somewhat opinionated discussion was added by [Max Held](https://www.maxheld.de).
+Your mileage may vary.
+:::
+
+Test doubles such as mocks or fakes can help you isolate your tests from the state of external HTTP API.
+This can be a good thing, because it allows you to test your code against a rare responses, or when the API is not available.
+
+Unfortunately, mocking and faking add their own complexity to a project.
+Respectively, you have to carry expected responses or even (rudimentary) API behavior in your code base.
+You now have two relationships to keep in mind: your code must pass your mocked/faked tests, and your mocks/fakes must correspond to the live API.
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
+Conversely, if you find yourself requiring these test doubles in a lot of places, you may want to [consider a refactoring](https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a).
```

Comment:
Instead of an external link or as a complement to it I'd add an example or maybe if you add the mock code earlier it is enough.

Reply:
I pointed to the below example here now, but again, would benefit from the full integration with the examples you suggested.